core/src/main/java/jenkins/SlaveToMasterFileCallable.java

```diff
@@ -1,18 +1,46 @@
 package jenkins;
 
 import hudson.FilePath.FileCallable;
+import hudson.Main;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 import jenkins.security.Roles;
+import jenkins.util.JenkinsJVM;
 import org.jenkinsci.remoting.RoleChecker;
 
 /**
  * {@link FileCallable}s that can be executed on the master, sent by the agent.
  * Note that any serializable fields must either be defined in your plugin or included in the stock JEP-200 whitelist.
+ * Additionally, this callable can be called with any {@link hudson.FilePath}, it is your responsibility to validate it
+ * in {@link #invoke(java.io.File, hudson.remoting.VirtualChannel)}.
  * @since 1.587 / 1.580.1
+ * @deprecated Use {@link jenkins.security.SlaveToMasterCallable} instead (and only if you really have to), and think
+ * carefully about the <a href="https://www.jenkins.io/doc/developer/security/remoting-callables/">security implications</a>.
+ *
+ * @see jenkins.security.SlaveToMasterCallable
+ * @see org.jenkinsci.remoting.RoleSensitive
  */
+@Deprecated
 public abstract class SlaveToMasterFileCallable<T> implements FileCallable<T> {
+
+    public static final Logger LOGGER = Logger.getLogger(SlaveToMasterFileCallable.class.getName());
+
     @Override
     public void checkRoles(RoleChecker checker) throws SecurityException {
+        warnOnController();
         checker.check(this, Roles.MASTER);
     }
+
+    protected Object readResolve() {
+        warnOnController();
+        return this;
+    }
+
+    private void warnOnController() {
+        if (JenkinsJVM.isJenkinsJVM() && (Main.isUnitTest || Main.isDevelopmentMode)) { // No point in spamming admins who cannot do anything
```

Comment:
```suggestion
        if (JenkinsJVM.isJenkinsJVM()) {
```
They can do something: report a bug; uninstall the plugin; even adopt it. We will never catch long-tail usages based on absentee maintainers paying attention.

Reply:
I think we can progressively restrict it further, no need to start with making plugins fail outright.

Reply:
> long-tail usages

S2MFCs can be statically identified, at least in Jenkins project and CloudBees hosted plugins. This isn't `FilePath` all over.

Reply:
> making plugins fail outright

Was this for https://github.com/jenkinsci/jenkins/pull/5884#discussion_r743899465?

Reply:
Applies to both in a way? The possible progression is probably:

1. Deprecate only
2. Deprecate and be noisy during tests (this)
3. Deprecate and be noisy
4. Deprecate and fail stuff

Reply:
Right, so I would recommend 3 or 4, but will defer to your judgment.

Reply:
I would prefer (3) but fine with (2)

Reply:
> I would prefer (3) but fine with (2)

Ditto, for the reason given in [this comment](https://github.com/jenkinsci/jenkins/pull/5884/files#r743898577).